Synertek Systems Sym-1 files

WayneParham opened the pull request:
These are the machine-specific files to support the Synertek Systems Sym-1.

groessler commented on cfg/sym1.cfg:
```diff
@@ -0,0 +1,46 @@
+}
+
+MEMORY {
+	ZP:		start = $0000,  size = $00F7,                      define = yes,   file = %O;
```
Please don't ise TAB chars.

I'm seeing TAB chars on these files of you:

cfg/sym1.cfg
cfg/sym1-4k.cfg
cfg/sym1-32k.cfg
samples/symIO.c
samples/tutorial/sym1/symIO.c
samples/tutorial/sym1/symNotepad.c
samples/tutorial/sym1/symDisplay.c
samples/symNotepad.c
samples/symDisplay.c

WayneParham replied:
I agree completely.  I hate TAB characters in source files 'cause it skews the display compared to lines with space chars only, making viewing ugly and difficult.

I'll modify those files to remove TABs and push them to my branch.

WayneParham replied:
--- TABs removed ---

oliverschmidt commented:
First of all thanks for your interest to make this substantial contribution to cc65 :-) However, please consider:

* We are in certain aspects pretty strict regarding code style - incl. aspects not even documented but only visible in the exsisting code.
* I'm in general not doing any PR-post-processing. In othe words, a PR must match the desired code style in order to be merged.

I'm emphazing this as I've already experienced a PR withdrawn after sigificant effort spent on both sides to adjust it with complaining me acting as "style police". So please acknowledge that you are willing to adjust things until they match the exsisting code base before we start to invest time.

WayneParham commented:
Of course, I understand.  Let me know if you see any style changes you want and I'll adjust the code in this pull request to match the style of code present in the repository.

To be honest, I tried to parrot the style I saw already.  So we should be pretty close.   And like I said, if we want to adjust style further to match closer - perhaps in ways I just didn't see - I'm happy to do that.  Just let me know.

oliverschmidt commented:
Great :-)

Unfortunately I'm currently (too) busy to provide useful, detailed feedback - but rest assured that I will do so asap.

WayneParham commented:
OK, I'll make sure to keep my branch current with cc65:master until you are able to review and merge the Synertek code.

oliverschmidt commented on samples/Makefile:
```diff
@@ -471,9 +489,10 @@ zip:
 
 mostlyclean:
-	@$(DEL) *.lbl *.map *.o *.s 2>$(NULLDEV)
+	@$(DEL) *.lbl *.map *.bin *.hex *.o *.s 2>$(NULLDEV)
```
- The ld65 binary output files have no special file extension, so there's no need to remove .bin files.
- There's no cc65 program generating .hex files so there's no need to remove them. 

So please remove this change again.

WayneParham replied:
**DONE**.  Please check and confirm.

oliverschmidt commented on samples/symDisplay.c:
```diff
@@ -0,0 +1,360 @@
+   puts( "\n\nEnjoy your day!\n\n" );
+
+   return;
+}
```
Please change this code to adhere to the cc65 C code style rules - see https://github.com/cc65/wiki/wiki/C-and-Assembly-sources-style-guide

WayneParham replied:
I've removed the 'return' line.  Honestly, if these programs were going to run on a computer with a filesystems, I would have always made them return something so I could test for error by the calling process.  But the Sym-1 is more like a microcontroller, and there's nothing that would catch the return value so it makes sense to have main() not return a value.

oliverschmidt commented on samples/symHello.c:
```diff
@@ -0,0 +1,39 @@
+   }
+
+   return;
+}
```
Please change this code to adhere to the cc65 C code style rules - see https://github.com/cc65/wiki/wiki/C-and-Assembly-sources-style-guide

WayneParham replied:
**DONE**.  Please check and confirm.

oliverschmidt commented on samples/symIO.c:
```diff
@@ -0,0 +1,170 @@
+   }
+
+   return;
+}
```
Please change this code to adhere to the cc65 C code style rules - see https://github.com/cc65/wiki/wiki/C-and-Assembly-sources-style-guide

WayneParham replied:
**DONE**.  Please check and confirm.

oliverschmidt commented on samples/symNotepad.c:
```diff
@@ -0,0 +1,199 @@
+
+   return;
+}
+
```
Please change this code to adhere to the cc65 C code style rules - see https://github.com/cc65/wiki/wiki/C-and-Assembly-sources-style-guide

WayneParham replied:
**DONE**.  Please check and confirm.

oliverschmidt commented on samples/symTiny.c:
```diff
@@ -0,0 +1,42 @@
+   }
+
+   return;
+}
```
Please change this code to adhere to the cc65 C code style rules - see https://github.com/cc65/wiki/wiki/C-and-Assembly-sources-style-guide

WayneParham replied:
**DONE**.  Please check and confirm.

oliverschmidt commented on samples/tutorial/sym1/build:
```diff
@@ -0,0 +1,36 @@
+
+
+fi
+
```
We don't have/want any batch file in this repository, so please remove this file.

In case you want to replace this file with Makefile logic, please be aware that bin2hex is neither part of the cc65 toolchain nor is it listed in https://www.gnu.org/software/make/manual/make.html#Utilities-in-Makefiles so you can't just call it.

WayneParham replied:
**REMOVED**

oliverschmidt commented on samples/tutorial/sym1/build_32k:
```diff
@@ -0,0 +1,36 @@
+
+
+fi
+
```
Same as above.

WayneParham replied:
**REMOVED**

oliverschmidt commented on samples/tutorial/sym1/build_4k:
```diff
@@ -0,0 +1,36 @@
+
+
+fi
+
```
Same as above.

WayneParham replied:
**REMOVED**

oliverschmidt commented on samples/tutorial/sym1/clean:
```diff
@@ -0,0 +1,14 @@
+   echo "--- Cleaned  $1 ---"
+
+fi
+
```
We don't have/want any batch file in this repository, so please remove this file.

WayneParham replied:
**REMOVED**

oliverschmidt commented on samples/tutorial/sym1/readme.txt:
```diff
@@ -0,0 +1,11 @@
+
+Usage: build <program>   (don't include the .c extension)
+       clean <program>   (removes intermediate and output files)
+
```
Without the batch files, there's no point in this file.

WayneParham replied:
**REMOVED**

oliverschmidt commented on samples/tutorial/sym1/symDisplay.c:
```diff
@@ -0,0 +1,360 @@
+   puts( "\n\nEnjoy your day!\n\n" );
+
+   return;
+}
```
Please change this code to adhere to the cc65 C code style rules - see https://github.com/cc65/wiki/wiki/C-and-Assembly-sources-style-guide

WayneParham replied:
**DONE**.  Please check and confirm.

oliverschmidt commented on samples/tutorial/sym1/symDisplay.c:
```diff
@@ -0,0 +1,360 @@
+// --------------------------------------------------------------------------
```
We don't want any new programs in the tutorial. Please remove it or move it to `targettest/sym1`.

WayneParham replied:
Moved to targettest.

oliverschmidt commented on samples/tutorial/sym1/symHello.c:
```diff
@@ -0,0 +1,39 @@
+// --------------------------------------------------------------------------
```
We don't want any new programs in the tutorial. Please remove it or move it to `targettest/sym1`.

WayneParham replied:
Moved to targettest/sym1.

oliverschmidt commented on samples/tutorial/sym1/symIO.c:
```diff
@@ -0,0 +1,170 @@
+// --------------------------------------------------------------------------
```
We don't want any new programs in the tutorial. Please remove it or move it to `targettest/sym1`.

WayneParham replied:
**FIXED**  Please check and confirm.

oliverschmidt commented on samples/tutorial/sym1/symNotepad.c:
```diff
@@ -0,0 +1,199 @@
+// --------------------------------------------------------------------------
```
We don't want any new programs in the tutorial. Please remove it or move it to `targettest/sym1`.

WayneParham replied:
**FIXED**  Please check and confirm.

oliverschmidt commented on samples/tutorial/sym1/symTiny.c:
```diff
@@ -0,0 +1,42 @@
+// --------------------------------------------------------------------------
```
We don't want any new programs in the tutorial. Please remove it or move it to `targettest/sym1`.

WayneParham replied:
**FIXED**   Please check and confirm.

oliverschmidt commented on src/ca65/main.c:
```diff
@@ -334,6 +334,9 @@ static void SetSys (const char* Sys)
             break;
 
+        case TGT_SYM1:
+            break;
```
Please don't get me wrong, but have you even looked at the case'es above? Why didn't you add

```
NewSymbol ("__SYM1__", 1);
```

WayneParham replied:
**FIXED**.  I can't believe I missed that.  Please check and confirm.

oliverschmidt commented on src/cc65/main.c:
```diff
@@ -290,6 +290,9 @@ static void SetSys (const char* Sys)
             break;
 
+        case TGT_SYM1:
+            break;
```
See above.

WayneParham replied:
**FIXED**.  Please check and confirm.

oliverschmidt reviewed (changes requested):
1. Sorry for taking so long to do this review. Thanks for your patience.

2. Apart from my comments it seems to me that something like https://github.com/cc65/cc65/blob/master/libsrc/supervision/ctype.s is missing.

3. Content for https://cc65.github.io/doc/ca65.html#s14 and https://cc65.github.io/doc/cc65.html#s6 is missing.

polluks commented:
A void parameter function cannot be a fastcall at all.

WayneParham commented:
> 1. Sorry for taking so long to do this review. Thanks for your patience.

Thanks so much for taking the time to do such an extensive code review!  I have made a first round of changes, and will be making more as time permits.

A quick aside:  When I found this cc65 compiler, it seemed to me that it would be wonderful to include the KIM-1 since it was the first 6502 system and the SYM-1 since it was the second.

I didn't feel comfortable making code for the KIM-1 because I don't have one to test on.  But I do have a SYM-1.  And what's nice is the SYM-1 is very similar to the KIM-1, even having monitor entry points and system memory locations with the same mnemonics in the assembler listings and programming manuals.  So a person with a KIM-1 can probably use this SYM-1 target as a staring point.

Anyway, thanks so much for taking the time to get this right.  We'll work together to make it fit properly in the cc65 style.

> 2. Apart from my comments it seems to me that something like https://github.com/cc65/cc65/blob/master/libsrc/supervision/ctype.s is missing.

**FIXED**

> 3. Content for https://cc65.github.io/doc/ca65.html#s14 and https://cc65.github.io/doc/cc65.html#s6 is missing.

**FIXED**

WayneParham commented on asminc/sym1.inc:
```diff
@@ -0,0 +1,186 @@
+L8C78    := $8C78     ; Link to tape
+DUMPT    := $8E87     ; Dump memory to tape
+LOADT    := $8C78     ; Load memory from tape
+TAPEMODE := $00FD     ; Top bit on for high-speed
```
**FIXED**   Please check and confirm.

WayneParham commented on cfg/sym1-32k.cfg:
```diff
@@ -0,0 +1,46 @@
+#
+# ld65 --config sym1-32k.cfg -o <prog>.bin <prog>.o
+
+FEATURES {
```
I copied the style of c16.cfg/c16-32k.cfg, supervision-16k.cfg/supervision-64k.cfg/supervision-128k.cfg and vic20.cfg/vic20-32k.cfg.

The sym1-4k.cfg configuration is for a system with only the onboard 4Kb RAM and the sym1-32k.cfg configuration is for a system with an expansion memory board.

Isn't this appropriate?  It appears that is the same reasoning employed when providing the multiple config files for the c16, supervision and vic20 systems.

So are you sure you want me to delete one of the configuration files?

WayneParham commented on cfg/sym1-32k.cfg:
```diff
@@ -0,0 +1,46 @@
+    CONDES:    segment        = STARTUP,
+               type           = destructor,
+               label          = __DESTRUCTOR_TABLE__,
+               count          = __DESTRUCTOR_COUNT__;
```
**FIXED**  Please check and confirm.

WayneParham commented on cfg/sym1-32k.cfg:
```diff
@@ -0,0 +1,46 @@
+        ONCE:           load = RAM,     type = ro,                         define = yes;
+        DATA:           load = RAM,     type = rw,                         define = yes;
+        BSS:            load = RAM,     type = bss,                        define = yes;
+}
```
**FIXED**  Please check and confirm.

WayneParham commented on cfg/sym1-32k.cfg:
```diff
@@ -0,0 +1,46 @@
+}
+
+MEMORY {
+        ZP:             start = $0000,  size = $00F7,                      define = yes,   file = %O;
```
**FIXED**  Please check and confirm.

WayneParham commented on cfg/sym1-4k.cfg:
```diff
@@ -0,0 +1,46 @@
+        DATA:           load = RAM,     type = rw,                         define = yes;
+        BSS:            load = RAM,     type = bss,                        define = yes;
+}
+
```
**FIXED**.  Please check and confirm.

WayneParham commented on cfg/sym1.cfg:
```diff
@@ -0,0 +1,46 @@
+# sym1-4k.cfg   (4k)
+#
+# for Sym-1 with 4kb RAM
```
The sym1.cfg file is a copy of the sym1-4k.cfg file.  My thought was it would be a default configuration.

WayneParham commented on cfg/sym1.cfg:
```diff
@@ -0,0 +1,46 @@
+        DATA:           load = RAM,     type = rw,                         define = yes;
+        BSS:            load = RAM,     type = bss,                        define = yes;
+}
+
```
**FIXED**.  Please check and confirm.

WayneParham commented on doc/sym1.sgml:
```diff
@@ -0,0 +1,136 @@
+
+<sect>Binary format<p>
+
+The standard binary output format generated by the linker for the Sym-1 target is a 4 kbyte machine language program.  It is, of course, possible to change this behavior by using one of the different linker configs.
```
I think so.  Anyone that uses a SYM-1 must be familiar with machine code.  However, it might make sense here to make mention of the bin2hex program, external to cc65.  That gives a person a way to enter the program into the monitor.

polluks replied:
Better use `xxd` which is part of Linux and macOS.

WayneParham replied:
The Sym-1 is closely related to embedded systems, where the BIN file extension is popular.  The HEX format is also popular, so embedded developers often translate between BIN and HEX files using a converter.

oliverschmidt replied:
Maybe there's a misunderstanding: The question is which "tool" the user is supposed to run the binary on. Is it a real machine? Is it an emulator? Both? Based on that decision it is desirable to provide hints (aka explain very briefly) on the workflow from the ld65 output file to the running program.

WayneParham replied:
Oh, I see now Stefan and Oliver.  You're right - It would be useful to explain a little more about what to do with the BIN file and how to get it onto the Sym-1.  Or at least describe what the files are and how they relate to the Sym-1, allowing the user to choose how they want to transfer the file.

I'll write a sentence or two about that and update the sym1 document soon.

WayneParham commented on doc/sym1.sgml:
```diff
@@ -0,0 +1,136 @@
+
+<sect1>Hardware access<p>
+
+The following pseudo variables declared in the <tt/sym1.inc/ include file allow access to hardware located in the address space.  See the include file for more information.
```
I wrote, _"See the include file for more information."_  Would you prefer that I list each of them in this document?

oliverschmidt replied:
No, but at least according to my knowledge of English it's not correct to write `the following` if nothing actualls follows **HERE** (not somewhere else).

WayneParham replied:
You're right.  I've updated the Sym-1 documentation page accordingly.

WayneParham commented on doc/sym1.sgml:
```diff
@@ -0,0 +1,136 @@
+
+<sect2>Limited memory applications<p>
+
+As stated earlier, there are config files for 4kb and 32kb systems.  If you have 32kb RAM, then you will probably want to use the sym1_32k configuration, but if not - if you are using the sym1_32k configuration - then you may want to use functions like getchar, putchar, gets and puts rather than printf.  Printf requires about 1kb because it needs to know how to process all the format specifiers.
```
**FIXED**.  Good catch.  The sentence about the differences between the 4KB and 32KB configuration files was wrong.  I also changed all references to 4kb and 32kb to 4KB and 32KB.

WayneParham commented on doc/sym1.sgml:
```diff
@@ -0,0 +1,136 @@
+
+<sect3>Sample programs<p>
+
+All the samples will run on the &quot;stock&quot; 4k Sym-1, except for symio.c, which requires 8k.  Information on building and running it is in the samples/tutorial/sym1 directory.
```
The program symIO.c is a sample program.  It compiles down to just under 8KB.  Might not have been a relevant comment, but that's what I was saying there.

oliverschmidt replied:
The only way I can interpret `which requires 8k` is this: A machine with 8 kB can execute this program. But if someone has a machine with exactly 8 kB: Is there a way to make the program run? As far as I understand your cfg files, there's no linker config file allowing to do so.

WayneParham replied:
You're right. I've updated the Sym-1 documentation page accordingly.

WayneParham commented on include/symio.h:
```diff
@@ -0,0 +1,148 @@
+//
+// I/O primitives for Sym-1
+//
+// Wayne Parham
```
**FIXED**.  Please check and confirm.

WayneParham commented on include/symio.h:
```diff
@@ -0,0 +1,148 @@
+// I/O primitives for Sym-1
+//
+// Wayne Parham
+
```
**FIXED**.  Please check and confirm.

WayneParham commented on include/symio.h:
```diff
@@ -0,0 +1,148 @@
+
+#ifndef _SYMIO_H
+#define _SYMIO_H
+
```
**FIXED**.  Please check and confirm.

WayneParham commented on include/symio.h:
```diff
@@ -0,0 +1,148 @@
+
+#include <stddef.h>
+#include <stdarg.h>
+
```
**FIXED**.  Please check and confirm.

WayneParham commented on include/symio.h:
```diff
@@ -0,0 +1,148 @@
+int   __fastcall__   set_DDR3B (int);        // Set data direction register 3B           (U29)
+int   __fastcall__   get_DDR3B (void);       // Get value of data direction register 3B
+int   __fastcall__   set_IOR3B (int);        // Set I/O register 3B
+int   __fastcall__   get_IOR3B (void);       // Get value of I/O register 3B
```
**FIXED**.  Please check and confirm.

WayneParham commented on include/symio.h:
```diff
@@ -0,0 +1,148 @@
+int   __fastcall__   vscanf (const char* format, va_list ap);
+int   __fastcall__   vsscanf (const char* s, const char* format, va_list ap);
+
+#endif
```
**FIXED**.  Please check and confirm.

WayneParham commented on libsrc/sym1/beep.s:
```diff
@@ -0,0 +1,24 @@
+; Wayne Parham
+;
+; wayne@parhamdata.com
+; ---------------------------------------------------------------------------
```
**FIXED**.  Please check and confirm.

WayneParham commented on libsrc/sym1/beep.s:
```diff
@@ -0,0 +1,24 @@
+.segment        "CODE"
+
+.proc _beep:    near
+; ---------------------------------------------------------------------------
```
**FIXED**.  Please check and confirm.

WayneParham commented on libsrc/sym1/beep.s:
```diff
@@ -0,0 +1,24 @@
+        lda     #$00             ;
+        ldx     #$00             ;
+        rts                      ;  Return 0000
+; ---------------------------------------------------------------------------
```
**FIXED**.  Please check and confirm.

WayneParham commented on libsrc/sym1/beep.s:
```diff
@@ -0,0 +1,24 @@
+        jsr     BEEP             ;  Beep
+        lda     #$00             ;
+        ldx     #$00             ;
+        rts                      ;  Return 0000
```
**FIXED**.  Please check and confirm.

WayneParham commented on include/symio.h:
```diff
@@ -0,0 +1,148 @@
+#include <stdarg.h>
+
+int   __fastcall__   beep (void);            // Beep sound
+int   __fastcall__   set_D0 (char);          // Set display digit 0
```
**FIXED**.  Please check and confirm.

WayneParham commented on libsrc/Makefile:
```diff
@@ -148,18 +149,26 @@ GEOSDIRS = common      \
+# MACHINE set independently of TARGET lets us easily define new targets
+# without  changing target.h and target.c in  cc65/src/common.   Useful
+# for initial testing of ports to new systems.
+
```
**FIXED**.  Please check and confirm.

WayneParham commented on libsrc/Makefile:
```diff
@@ -173,16 +182,26 @@ ifeq ($(TARGET),$(filter $(TARGET),$(GEOS)))
+             serial   \
+             tgi      \
+             zlib
+endif
```
**FIXED**.  Please check and confirm.

WayneParham commented on include/symio.h:
```diff
@@ -0,0 +1,148 @@
+
+int   __fastcall__   beep (void);            // Beep sound
+int   __fastcall__   set_D0 (char);          // Set display digit 0
+int   __fastcall__   get_D0 (void);          // Get value of display digit 0
```
**FIXED**.  Please check and confirm.

WayneParham commented on libsrc/sym1/bitio.s:
```diff
@@ -0,0 +1,230 @@
+
+.include        "sym1.inc"
+
+.export         _set_DDR1A, _get_DDR1A, _set_IOR1A, _get_IOR1A
+.export         _set_DDR1B, _get_DDR1B, _set_IOR1B, _get_IOR1B
+.export         _set_DDR2A, _get_DDR2A, _set_IOR2A, _get_IOR2A
+.export         _set_DDR2B, _get_DDR2B, _set_IOR2B, _get_IOR2B
+.export         _set_DDR3A, _get_DDR3A, _set_IOR3A, _get_IOR3A
+.export         _set_DDR3B, _get_DDR3B, _set_IOR3B, _get_IOR3B
```
I like using getters and setters.  I use PEEK and POKE in BASIC, but I prefer getters and setters in C.  More "C-Like."  :-)

polluks replied:
Rather "C++ overhead" IMHO.

oliverschmidt replied:
You got me wrong. I propose that you implement your getters and setters as macros using PEEK and POKE internally.

oliverschmidt replied:
Independently from the question how the getters and setters are implemented my first question is not answered yet: Is it really desirable to have many individual getters and setter? I imagine that it might very well be preferable to have just one getter and one setter with the register to get/set being a parameter. And to have a `#define` for every register name:

```
set(DDR1A, val);
val = get(DDR1A);
```

WayneParham replied:
Good suggestions.  I've modified sym1.h to use structs instead of getters and setters.

WayneParham commented on include/symio.h:
```diff
@@ -0,0 +1,148 @@
+// symio.h
```
**FIXED**.  Please check and confirm.

WayneParham commented on libsrc/sym1/bitio.s:
```diff
@@ -0,0 +1,230 @@
+        sta     DDR1A            ;  Write data direction register for port 1A
+        lda     #$00             ;
+        ldx     #$00             ;
+        rts                      ;  Return 0000
```
**FIXED**.  Please check and confirm.

WayneParham commented on libsrc/sym1/getchar.s:
```diff
@@ -0,0 +1,21 @@
+
+.segment        "CODE"
+
+.proc _getchar: near
```
**FIXED**.  Please check and confirm.

WayneParham commented on libsrc/sym1/putchar.s:
```diff
@@ -0,0 +1,23 @@
+
+.segment        "CODE"
+   
+.proc _putchar: near
```
**FIXED**.  Please check and confirm.

WayneParham commented on libsrc/sym1/libref.s:
```diff
@@ -0,0 +1,3 @@
+
+        .import         _exit
```
**FIXED**.  Please check and confirm.

WayneParham commented on libsrc/sym1/tapeio.s:
```diff
@@ -0,0 +1,58 @@
+        jmp     done
+error:  ldx     #$00
+        lda     #$FF             ; or 00FF if not
+done:   rts
```
Great idea.  **FIXED**.  Please check and confirm.

WayneParham commented on libsrc/sym1/tapeio.s:
```diff
@@ -0,0 +1,58 @@
+        jmp     done
+error:  ldx     #$00
+        lda     #$FF             ; or 00FF if not
+done:   rts
```
Again, great idea.  **FIXED**.  Please check and confirm.

WayneParham commented on samples/Makefile:
```diff
@@ -8,6 +8,10 @@
 # var. to build for another target system.
 SYS ?= c64
 
+# New targets can define MACHINE separately
+# from target if needed.
+MACHINE=$(SYS)
+
```
**FIXED**.  Please check and confirm.

WayneParham commented on samples/Makefile:
```diff
@@ -135,26 +139,31 @@ LDFLAGS_tgidemo_atarixl    = --start-addr 0x4000
 ifeq ($(SYS),vic20)
 	$(LD) $(LDFLAGS_$(@F)_$(SYS)) $(LDFLAGS) -o $@ -C vic20-32k.cfg -m $@.map $^ $(SYS).lib
+else ifeq ($(SYS),sym1)
+	$(LD) $(LDFLAGS_$(@F)_$(SYS)) $(LDFLAGS) -o $@.bin -C sym1.cfg -m $@.map $^ $(SYS).lib
```
I liked the "bin" extension but removed this line on request.

WayneParham commented on samples/Makefile:
```diff
@@ -267,10 +276,12 @@ EXELIST_gamate =   \
         hello
 
 EXELIST_geos-cbm = \
+        helloworld \
```
**Removed**

WayneParham commented on samples/Makefile:
```diff
@@ -267,10 +276,12 @@ EXELIST_gamate =   \
         diodemo
 
 EXELIST_geos-apple = \
+        helloworld   \
```
**Removed**

WayneParham commented on samples/helloworld.c:
```diff
@@ -0,0 +1,9 @@
+
+void main (void)
+{
+   printf( "Hello World!\n" );
+    return;
```
**Removed**

WayneParham commented on samples/Makefile:
```diff
@@ -309,6 +320,12 @@ EXELIST_sim65c02 = $(EXELIST_sim6502)
+        helloworld \
+        symHello   \
+        symTiny    \
+        symDisplay
```
I've updated the samples/Makefile to remove these entries and only included the traditional helloworld.c program.  I've also moved the Sym-specific programs into the targettest/sym1 directory.

WayneParham commented on samples/Makefile:
```diff
@@ -330,12 +347,13 @@ ifndef EXELIST_$(SYS)
 endif
 
-define SUBDIR_recipe
+define  SUBDIR_recipe
```
**DONE**.  Please check and confirm.

WayneParham commented on samples/Makefile:
```diff
@@ -330,12 +347,13 @@ ifndef EXELIST_$(SYS)
 
 endef # SUBDIR_recipe
 
+
```
**DONE**.  Please check and confirm.

WayneParham commented on samples/Makefile:
```diff
@@ -471,9 +489,10 @@ zip:
 	@$(DEL) multdemo.? ovrldemo.? 2>$(NULLDEV)
-	$(foreach dir,$(DIRLIST),$(SUBDIR_recipe))
+	@$(foreach dir,$(DIRLIST),$(SUBDIR_recipe))
+
```
**DONE**.  Please check and confirm.

WayneParham commented:
OK, I think I've addressed all the change requests.  And the code builds, so that's good.  But I need to test all the functions this week before we can merge.  I need to test the tape functions, since we've modified that code.  I also need to test the digital I/O code.  And even the basic terminal I/O - the putchar(), getchar() and printf() functions need to be verified.  I want to make sure we don't see any weird behaviors like double echoes, linefeeds where they shouldn't be (or lack of linefeeds where they are expected), etc.

I'm pretty busy this week so it may be a day or three before I report back.

WayneParham commented:
I added a Makefile for the targettest/sym1 directory.  I like this much better than the build scripts I had in the tutorial/sym1 directory.  The Makefile is more inline with the rest of the cc65 distribution.

I also was able to take advantage of the two config files, since the first three programs fit into 4KB but the last two do not.  The last two require the 32KB config file.  This is neatly shown in the Makefile.

I was also able to make the output files have the BIN extension.  I like this for the Sym-1 because the BIN extension is so popular in the embedded world, and that's really the audience for the Sym-1 back when it was created.  It was an evaluation board, but it leaned heavily towards the embedded market.  Most embedded system developers are familiar with BIN files, so this is a natural filename choice.

WayneParham commented:
- The "helloworld" program in the samples directory has been tested.  It works.

- But the getchar() function doesn't work properly with the Sym-1.  Programs using it echo the character twice.  They also do not strip off the top bit.  I had resolved this in my original code with a Sym-specific getchar() function in the libsrc\sym1 directory.

WayneParham commented:
- symTiny in the targettest/sym1 directory has been tested.  It works.

WayneParham commented:
- symHello in the targettest/sym1 directory has been tested. It works.

WayneParham commented:
- symDisplay in the targettest/sym1 directory has been tested. It works.

oliverschmidt commented on libsrc/sym1/tapeio.s:
```diff
@@ -0,0 +1,47 @@
+.proc _loadt:   near
+
+        sta     P1L              ; Tape record ID to P1L
+        ldx     #$00
```
When this function has an `int` parameter, meaning 16 bit and you set `P1` which seems to be a 16 bit value, why do you then trash the high byte by setting X to zero?

WayneParham replied:
Updated to cast "ID" as unsigned char rather than int.

oliverschmidt commented on libsrc/sym1/tapeio.s:
```diff
@@ -0,0 +1,47 @@
+        stx     P2H
+        jsr     popax
+        sta     P1L              ; Tape Record ID
+        ldx     #$00
```
When this function has an `int` parameter, meaning 16 bit and you set `P1` which seems to be a 16 bit value, why do you then trash the high byte by setting X to zero?

WayneParham replied:
Updated to cast "ID" as unsigned char rather than int.

oliverschmidt commented:
A more general thought/idea:

The `bell()` function would probably better be replaced by the `send()` function taking care of the value $07 (allowing to write:

```
putchar('\a');
```

See e.g. https://en.wikipedia.org/wiki/Bell_character.

oliverschmidt commented on cfg/sym1-32k.cfg:
```diff
@@ -0,0 +1,46 @@
+}
+
+SEGMENTS {
+        ZEROPAGE:  load = ZP,  type = zp, define = yes;
+        STARTUP:   load = RAM, type = ro, define = yes;
+        CODE:      load = RAM, type = ro, define = yes;
+        RODATA:    load = RAM, type = ro, define = yes;
+        ONCE:      load = RAM, type = ro, define = yes;
+        DATA:      load = RAM, type = rw, define = yes;
```
Sorry for the nitpicking, but can you please align the `define`s with the one below?

WayneParham replied:
**Done**.  Please check and confirm.

oliverschmidt commented on include/sym1.h:
```diff
@@ -0,0 +1,154 @@
+unsigned char __fastcall__ get_D6 (void);    // Get value of memory byte to the right of display
+void __fastcall__ fdisp (void);              // Flash display
+
+int __fastcall__ loadt (int);                // Read from tape (id)
```
Is the tape ID really a signed 16 value?

WayneParham replied:
No, it's not.  It's a single byte, so unsigned char is probably better suited.  I'll change that.  I'll also look into some of the other suggestions you've made - fastcall, address pointers, etc. - but I won't comment on those right now.  By leaving them "open," I will be reminded that I have open action items.

oliverschmidt commented on include/sym1.h:
```diff
@@ -0,0 +1,154 @@
+void __fastcall__ fdisp (void);              // Flash display
+
+int __fastcall__ loadt (int);                // Read from tape (id)
+int __fastcall__ dumpt (int, int, int);      // Write to tape (id, start_addr, end_addr)
```
Addresses for an output function are `const void*`, not `int`, see e.g. https://github.com/cc65/cc65/blob/master/include/unistd.h#L77

WayneParham replied:
Updated tapeio to expect const void* rather than int for start_addr and end_addr.

oliverschmidt commented on libsrc/sym1/bitio.s:
```diff
@@ -0,0 +1,220 @@
+;
+; Wayne Parham (wayne@parhamdata.com)
+;
+; void set_DDR1A (int value);
```
The reason to have the C prototpye repeated in the asm source code is to allow to easily check to correct parameter passing and return value generation.

From that perspective the `__fastcall__` is an actually very important aspect at this very place. Only the `__fastcall__` measn that the function can presume it's parameter to be in AX on entry.

Therefore please add the `__fastcall__` for all relevant functions into the asm source codes.

WayneParham replied:
Added __fastcall__ in asm source code comments to match C function declarations.

oliverschmidt commented on samples/helloworld.c:
```diff
@@ -0,0 +1,8 @@
+
+#include <stdio.h>
+
+int main() {
```
We don't want "old fashioned" C. Please use `int main(void)`.

WayneParham replied:
Very well.  Change made.

oliverschmidt commented on samples/helloworld.c:
```diff
@@ -0,0 +1,8 @@
+#include <stdio.h>
+
+int main() {
+   printf("Hello, World!");
```
The "traditional" hello world program includes a line feed - see e.g. https://en.wikipedia.org/wiki/%22Hello,_World!%22_program

WayneParham replied:
Very well.  Change made.

WayneParham commented:
Thanks for all the feedback!  I'm really busy this week, but I'll address each comment as time permits.

I also have to test a couple more test programs - symIO and symNotepad - and perhaps fix issues created by the recent changes.

So I'll reply back as soon as possible.  I may address some of the comments and leave others for later, and if I do that it is just because of time constraints.  I will get to each of them as soon as possible.

greg-king5 reviewed (commented):
We don't need another "Hello World!" program.  "samples/helloworld.c" is too similar to the code in "samples/tutorial/".  There's no advantage in having both of them.  I think that "samples/helloworld.c" should be removed.

WayneParham commented:
Greg, Stefan, that's a great idea using a struct to define the VIA.  I don't know why I missed that before.  I just have getters and setters on my mind, I suppose, 'cause I use them in a lot of code I do.  I'll use that approach instead of the getters and setters, and I'll re-write the symIO.c test program to use Stefan's 6522 struct.

I have a very busy week so my re-work and my responses to this code review will come in sporadically for a few days.  But I'll get to it as soon as possible.

WayneParham commented on libsrc/sym1/display.s:
```diff
@@ -0,0 +1,154 @@
+.export         _fdisp,  _set_D0, _get_D0
+.export         _set_D1, _get_D1, _set_D2, _get_D2
+.export         _set_D3, _get_D3, _set_D4, _get_D4
+.export         _set_D5, _get_D5, _set_D6, _get_D6
```
Good idea.  I've removed the getters and setters and moved to a struct approach as suggested above.

WayneParham commented:
Sorry for the delay in writing back.  I had a busy week, so I had to make the requested updates a little at a time.

But I think I have now addressed every one of the requested changes.

I must admit that I thought my initial code was good.  I was really happy to have a Sym-1 target for cc65, and I was happy with it.

But I also must admit that I love the changes you guys have suggested, and I think the code was improved by including them.

Thanks!

oliverschmidt commented on targettest/sym1/symDisplay.c:
```diff
@@ -0,0 +1,356 @@
+   }
+
+   puts( "\n\nEnjoy your day!\n\n" );
+}
```
Please adhere to the cc65 code style in general, as given by the two excamples above.

WayneParham replied:
Updated.  Please check and confirm.

oliverschmidt replied:
Sorry, for being not precise enough.

I wasn't talking about main() returning a value. It's perfectly fine that main() doesn't return a value for programs specific for a target where there's no ue for the return value of main().

I'm (still) talking about source code style. In cc65 it isn't `puts( "\n\nEnjoy your day!\n\n" );`. It's `puts ("\n\nEnjoy your day!\n\n");` .

WayneParham replied:
Oh, I see now.  OK, I'll make changes to my source files that reflect that sort of whitespace convention.

As an aside, I used to write code like that.  Actually, I did for many years.  But on the latest project I've been worlking on, the style is like what you saw.  So it has become a habit for the past few years.

Anyway, I'll go modify all the targettest samples to use the style you have described.

I'm going to leave leave the 'int main' and 'return int' convention.  It is sort of six one way and a half-dozen the other, since there is no operating system to capture the return code.  But it doesn't hurt anything either.  The Sym-1 monitor program will have a nice clean zero in the accumulator.  :-)

oliverschmidt commented on targettest/sym1/symHello.c:
```diff
@@ -0,0 +1,37 @@
+      for( d = 0; d < 10 ; d++ ) {
+      }
+   }
+}
```
Please adhere to the cc65 code style, see symDisplay.c.

WayneParham replied:
Updated.  Please check and confirm.

oliverschmidt replied:
Please adhere to the cc65 code style, see symDisplay.c.

WayneParham replied:
Updated.  Please check and confirm.

oliverschmidt commented on targettest/sym1/symIO.c:
```diff
@@ -0,0 +1,168 @@
+         }
+      }
+   }
+}
```
Please adhere to the cc65 code style, see symDisplay.c.

WayneParham replied:
Updated.  Please check and confirm.

oliverschmidt replied:
Please adhere to the cc65 code style, see symDisplay.c.

oliverschmidt commented on targettest/sym1/symNotepad.c:
```diff
@@ -0,0 +1,191 @@
+   free( buffer );
+
+   puts( "\nEnjoy your day!\n" );
+}
```
Please adhere to the cc65 code style, see symDisplay.c.

WayneParham replied:
Updated.  Please check and confirm.

oliverschmidt commented on targettest/sym1/symNotepad.c:
```diff
@@ -0,0 +1,191 @@
+
+   puts( "\nEnjoy your day!\n" );
+}
+
```
Remove this blank line.

WayneParham replied:
Updated.  Please check and confirm.

oliverschmidt commented on targettest/sym1/symTiny.c:
```diff
@@ -0,0 +1,40 @@
+      for( d = 0; d < 10 ; d++ ) {
+      }
+   }
+}
```
Please adhere to the cc65 code style, see symDisplay.c.

WayneParham replied:
Updated.  Please check and confirm.

oliverschmidt replied:
Please adhere to the cc65 code style, see symDisplay.c.

WayneParham replied:
Fixed.  Please check and confirm.

oliverschmidt commented on targettest/sym1/symIO.c:
```diff
@@ -0,0 +1,168 @@
+   unsigned char ior3a =   0x00;
+   unsigned char ddr3b =   0x00;
+   unsigned char ior3b =   0x00;
+   unsigned char   val =   0x00;
+   int           going =   0x01;
+   int           instr =   0x01;
+   int               l =   0x00;
+   char*            vp =   0x00;
+   char        cmd[20] = { 0x00 };
```
Please don't right-align anything. Here, `=` is the item to align.

WayneParham replied:
Updated.  Please check and confirm.

oliverschmidt commented on targettest/sym1/symNotepad.c:
```diff
@@ -0,0 +1,191 @@
+#define     TAPIO_MAX_SIZE     0x0FFF
+
+void main(void) {
+   char  c                 =   0x00;
+   int   l                 =   0x00;
+   int   p                 =   0x00;
+   int   error             =   0x00;
+   int   running           =   0x01;
+   int   writing           =   0x01;
```
Don't align variable names of different types. Here, `=` is the item to align.

WayneParham replied:
Updated.  Please check and confirm.

oliverschmidt commented on targettest/sym1/symNotepad.c:
```diff
@@ -0,0 +1,191 @@
+   int   running           =   0x01;
+   int   writing           =   0x01;
+   int instruction_needed  =   0x01;
+   int   heap_size         =   0x00;
```
Don't align variable names of different types. Here, `=` is the item to align.

WayneParham replied:
Updated.  Please check and confirm.

oliverschmidt commented on targettest/sym1/symNotepad.c:
```diff
@@ -0,0 +1,191 @@
+#define     TAPIO_MAX_SIZE     0x0FFF
+
+void main(void) {
+   char  c                 =   0x00;
```
Please remove the multiple blanks after the `=`. Same on the following lines.

WayneParham replied:
Updated.  Please check and confirm.

oliverschmidt commented on targettest/sym1/symNotepad.c:
```diff
@@ -0,0 +1,191 @@
+#include <stdlib.h>
+#include <string.h>
+
+#define     TAPIO_ADDRESS      0xE000
+#define     TAPIO_MAX_SIZE     0x0FFF
```
Don't put content on fixed columns. Here:

```
#define TAPIO_ADDRESS  0xE000
#define TAPIO_MAX_SIZE 0x0FFF
```

WayneParham replied:
Updated.  Please check and confirm.

oliverschmidt commented on include/sym1.h:
```diff
@@ -0,0 +1,133 @@
+
+
+
+void beep (void);                                                  // Beep sound
+void fdisp (void);                                                 // Flash display
+int __fastcall__ loadt (unsigned char);                            // Read from tape (id)
+int __fastcall__ dumpt (unsigned char, const void*, const void*);  // Write to tape (id, start_addr, end_addr)
```
Please adhere to the cc65 code style. Here:

```
void beep (void);
/* Beep sound */

void fdisp (void);
/* Flash display */

int __fastcall__ loadt (unsigned char id);
/* Read from tape */

int __fastcall__ dumpt (unsigned char id, const void* start_addr, const void* end_addr); 
/* Write to tape */
```

WayneParham replied:
Updated.  Please check and confirm.

oliverschmidt commented on doc/sym1.sgml:
```diff
@@ -0,0 +1,133 @@
+
+<sect2>Limited memory applications<p>
+
+As stated earlier, there are config files for 4KB and 32KB systems.  If you have 32KB RAM, then you will probably want to use the sym1-32k configuration, but if not - if you are using the sym1-4k configuration - then you may want to use functions like getchar, putchar, gets and puts rather than printf.  Printf requires about 1KB because it needs to know how to process all the format specifiers.
```
'getchar' and 'gets' have nothing to do with 'printf'.

WayneParham replied:
True.  I updated the comments to suggest using gets and puts rather than scanf and printf when memory space is tight.

oliverschmidt commented on doc/sym1.sgml:
```diff
@@ -0,0 +1,133 @@
+
+<sect>Binary format<p>
+
+The output format generated by the linker for the Sym-1 target is a raw binary BIN file, which is essentially a memory image.  You can convert this to a HEX file using BIN2HEX, which is a popular open-source conversion utility program.  A HEX file has ASCII representations of the hexadecimal byte values of the machine-language program.
```
Thanks for adding content here. However, I didn't think about an explanation of BIN and HEX files. I was thinking about what do to with the file generated by the linker in order to get it run. If getting a HEX file is part of the at process, it's enough to say something like "Use e.g. BIN2HEX to generate a HEX file." But the interesting part is what to do wit the HEX file!

WayneParham replied:
Yes, I understand.  I've added comments telling the user how to upload their HEX file onto the Sym-1.

oliverschmidt commented on targettest/sym1/Makefile:
```diff
@@ -0,0 +1,55 @@
+# Run 'make SYS=<target>'; or, set a SYS env.
+# var. to build for another target system.
+SYS ?= sym1
+
```
Remove these lines.

WayneParham replied:
Updated.  Please check and confirm.

oliverschmidt commented on targettest/sym1/Makefile:
```diff
@@ -0,0 +1,55 @@
+
+all: symHello.bin symTiny.bin symDisplay.bin symIO.bin symNotepad.bin
+
+symHello.bin: symHello.c
+	$(CL) -t $(SYS) -C sym1-4k.cfg -O -o symHello.bin symHello.c
+
+symTiny.bin: symTiny.c
+	$(CL) -t $(SYS) -C sym1-4k.cfg -O -o symTiny.bin symTiny.c
+
+symDisplay.bin: symDisplay.c
+	$(CL) -t $(SYS) -C sym1-4k.cfg -O -o symDisplay.bin symDisplay.c
+
+symIO.bin: symIO.c
+	$(CL) -t $(SYS) -C sym1-32k.cfg -O -o symIO.bin symIO.c
+
+symNotepad.bin: symNotepad.c
+	$(CL) -t $(SYS) -C sym1-32k.cfg -O -o symNotepad.bin symNotepad.c
```
Replace `$(SYS)` with `sym1`.

WayneParham replied:
Updated.  Please check and confirm.

oliverschmidt commented on samples/helloworld.c:
```diff
@@ -0,0 +1,8 @@
+int main(void) {
+   printf("Hello, World!\n");
+   return 0;
+}
```
I follow @greg-king5's argument in https://github.com/cc65/cc65/pull/1500#pullrequestreview-678114487 - so please remove this file (and your change to samples/Makefile).

WayneParham replied:
I have removed helloworld.c from the samples directory and adjusted the Makefile accordingly.

WayneParham commented on include/sym1.h:
```diff
@@ -0,0 +1,133 @@
+//
+// I/O primitives for Sym-1
+//
+// Wayne Parham
```
Updated.  Please check and confirm.

WayneParham commented on libsrc/sym1/read.s:
```diff
@@ -0,0 +1,53 @@
+;
+; Wayne Parham (wayne@parhamdata.com)
+;
+; int read (int fd, void* buf, unsigned count);
```
Updated.  Please check and confirm.

oliverschmidt replied:
Sorry, I don't see a change.

WayneParham replied:
You're right.  I failed to commit that change.  So sorry - It is now fixed.  Please check and confirm.

WayneParham commented on libsrc/sym1/write.s:
```diff
@@ -0,0 +1,52 @@
+;
+; Wayne Parham (wayne@parhamdata.com)
+;
+; int write (int fd, const void* buf, int count);
```
Updated.  Please check and confirm.

WayneParham commented on samples/Makefile:
```diff
@@ -154,6 +154,7 @@ endif
 # Lists of subdirectories
 
 # disasm depends on cpp
+
```
Updated.  Please check and confirm.

oliverschmidt replied:
Sorry, I can't see a change.

WayneParham replied:
You're right.  I failed to commit that change.  So sorry - It is now fixed.  Please check and confirm.

WayneParham commented on targettest/sym1/Makefile:
```diff
@@ -0,0 +1,55 @@
+all: symHello.bin symTiny.bin symDisplay.bin symIO.bin symNotepad.bin
+
+symHello.bin: symHello.c
+	$(CL) -t $(SYS) -C sym1-4k.cfg -O -o symHello.bin symHello.c
```
Updated.  Please check and confirm.

WayneParham commented on targettest/sym1/Makefile:
```diff
@@ -0,0 +1,55 @@
+	$(CL) -t $(SYS) -C sym1-4k.cfg -O -o symHello.bin symHello.c
+
+symTiny.bin: symTiny.c
+	$(CL) -t $(SYS) -C sym1-4k.cfg -O -o symTiny.bin symTiny.c
```
Updated.  Please check and confirm.

WayneParham commented on targettest/sym1/Makefile:
```diff
@@ -0,0 +1,55 @@
+	$(CL) -t $(SYS) -C sym1-4k.cfg -O -o symTiny.bin symTiny.c
+
+symDisplay.bin: symDisplay.c
+	$(CL) -t $(SYS) -C sym1-4k.cfg -O -o symDisplay.bin symDisplay.c
```
Updated.  Please check and confirm.

WayneParham commented:
I have made all the requested changes.  I think I understood you, Oliver, but if I've misunderstood you anywhere, please advise.

oliverschmidt commented on libsrc/sym1/crt0.s:
```diff
@@ -0,0 +1,59 @@
+          sta     TECHO
+          jsr     NACCES               ; Lock System RAM
+          rts                          ; Re-enter Sym-1 monitor
+
+
```
Please remove these empty lines.

WayneParham replied:
Fixed.  Please check and confirm.

oliverschmidt commented on libsrc/sym1/display.s:
```diff
@@ -0,0 +1,19 @@
+        rts
+
+.endproc
+
```
Please remove this empty line.

WayneParham replied:
Fixed.  Please check and confirm.

oliverschmidt commented on libsrc/sym1/read.s:
```diff
@@ -0,0 +1,53 @@
+        rts                      ; Return count                       
+
+.endproc
+
```
Please remove this empty line.

WayneParham replied:
Fixed.  Please check and confirm.

oliverschmidt commented on libsrc/sym1/tapeio.s:
```diff
@@ -0,0 +1,47 @@
+error:  jmp     return1          ; or 1 if not
+
+.endproc
+
```
Please remove this empty line.

WayneParham replied:
Fixed.  Please check and confirm.

oliverschmidt commented on libsrc/sym1/write.s:
```diff
@@ -0,0 +1,52 @@
+        rts                     ; Return count
+
+.endproc
+
```
Please remove this empty line.

WayneParham replied:
Fixed.  Please check and confirm.

WayneParham commented:
I believe I've made all the changes that were requested.

Please check them when time permits.  Let me know if I've missed any.

Fabrizio-Caruso commented:
@oliverschmidt could this PR be merged?

oliverschmidt commented:
This pull request seems to cause no regressions.

Fabrizio-Caruso commented:
@WayneParham how realistic and feasible would "display" output through an oscilloscope be? Have you by any chance, tried to implement it? Is there any emulator that supports the "oscilloscope display"?

polluks commented:
@Fabrizio-Caruso I suppose you have to implement this emulator feature yourself...

WayneParham commented:
That's a fantastic idea, Fabrizio.

The Sym-1 actually includes an oscilloscope output circuit, terminating on pin "R" of its edge connector "AA."

The software to map characters to O-Scope output isn't in the "monitor ROM," but Synertek did publish the assembly code in the Sym-1 reference manual.  So I could surely include that code in cc65, with a C function to expose it.

Nice call-out.

I simply overlooked it 'cause I always ran my Sym-1 with an ASCII terminal.

One caveat - I believe the "driver" software published in the Sym-1 Reference Manual is only capable of displaying alpha-numeric characters.  I do not know how many "special characters" like punctuation characters it is capable of displaying.  It could be that the character map will need to be updated to add characters for this driver software to be useful.  I'll need to play around with it to see.

Second caveat - I'm not entirely sure how the driver is designed to be called.  A quick scan of the assembly source shows it to be hard-coded to a 32-character line width.  It also appears that some of the input and output vectors - memory locations that point to hardware devices - are modified to point to the O-Scope, but then after that, the code in the monitor ROM that handles I/O can still be used.  This is the same way the Sym-1 handles selecting input and output to and from its front panel versus the ASCII port.  So I think anything we would do in cc65 would redirect output from the ASCII terminal to the O-Scope.

Might be able to modify that so it goes to both.  Probably could, probably easy enough.  I do that for the ASCII console output and the front-panel display in the symDisplay.c sample file.  It sends messages to both the screen and the front panel.  So it could probably work similarly with the O-Scope output.

Just thinkin' outloud.

I'll hook my scope up to my Sym-1 and play around with this in the next month or two.  Once I dial it in, I'll create a function for cc65 and test it out.  Then I'll create a pull request to provide that functionality.

Thanks again for bringing this to my attention.  It's a useful feature request.

Fabrizio-Caruso commented:
@WayneParham Great! Do you know of any emulator with oscilloscope output? It would be helpful to have one for you and for anyone who may want to code for it (for example me).

WayneParham commented:
Sorry, no.  I know there is a Sym-1 emulator available at http://6502.org.  There is also a Sym-1 (hardware) clone available at http://retro.hansotten.nl.  But I haven't worked with either or those.  I own a couple original Sym-1 computers and I test my code on them.

Fabrizio-Caruso commented:
@polluks I wish I were able to do it... I would be glad to test such feature on an emulator.
I will support the SYM-1 with oscilloscope video output in my universal game lib (Cross-Lib) if and when @WayneParham provides routines to display on the oscilloscope. I will gently ask him to test my games on the real hardware. An emulator would make all of this much simpler but it is not strictly necessary.

Fabrizio-Caruso commented:
@WayneParham Do you already provide any support for the ASCII terminal in your APIs? What do you mean by an ASCII terminal? A 80x24 text terminal capable of interpreting control codes? Which ones? ADM3/A? VT52? VT100? I could already include support for such terminals in my Cross-Lib universal game library and so also port all my games. 
Even if you do not support the ASCII terminal in the APIs, I could still support it if you tell me how I send characters (and control codes) to it.

WayneParham commented:
There is a "master console" on the Sym-1, and it is expected to be connected to an ASCII "dumb terminal."  The system didn't require a terminal - you could use it with its built-in keypad "front panel" - but it was much more useful as a general puprose computer if you connected a terminal.

I've used a variety of terminals on it over the years, mostly various Lear Siegler, DEC and Data General models.  Synertek made a terminal - seems like it was called the KTM-1 - and I owned one of those too.  Anyway, all of them had their own keymap and control sequence definitions.

And I think that's the key thing for what you're doing.  The Sym-1 didn't define any special sequences - nothing for cursor control or for primitive graphics - but the terminal certainly did.  Back in those days, many systems had configuration menus to allow selection of the terminal.  Lots of software and operating systems did that, to provide a way to connect a variety of terminals, each having their own escape sequences for cursor control and other such things.

Unix, for example, had a terminal definition file called "termcap" which mapped control sequences to standard functions on a variety of popular terminals.  So in unix of the day, you would select a termcap entry for each async port, and that would translate the desired action to the appropriate sequence for the terminal connected to that port.

Similarly, the popular word processor "WordStar" had a similar configuration utility.  It allowed the software to "know" what terminal was connected, so it could issue the right control sequence to manage the screen.  It was an early way to abstract functionality from the device hardware using what amounts to a lookup table.

I would suggest this approach as a possible solution for terminal support on any of the systems that allow a dumb terminal to be connected.  Since you don't have a built-in console - and therefore, no conio support - but instead have only an async port, the termnal connected cannot be known by any system software.  It has to be user-selectable by some form of configuration.  But you can always provide a sort of "termcap" in your software which allows selection of popular terminals.

WayneParham commented:
--- As I re-read your comment, if I understand you right, I think you already use this kind of approach in your software.  If so, just send the ASCII character(s) needed for cursor control for the terminal connected to the console.  The Sym-1 will send whatever data presented to stdout directly to the terminal.  So just printf() whatever escape sequences are needed and the terminal will respond.

Fabrizio-Caruso commented:
@WayneParham, indeed that is what I do for some targets including several CP/M based targets. I don't do it in the best possible way, yet (i.e., with a menu) but I do build ADM3/A, VT52 and VT100 versions for 80x24 terminals.
How many columns and rows do most of these terminals display? 80x24?
Do you know if any of the 3 escape/control standards mentioned above are understood by these terminals? 

I could release some games for your computer.

Are you aware of any SYM-1 emulator of emulating a terminal display?

WayneParham commented:
I haven't tried any of the hardware clones or software emulators for the Sym-1, so sorry, I'm no help to you there.

Most dumb terminals from the 1970s had 80 x 24 columns/rows.  Seems like I recall a terminal or maybe an early microcomputer with an 80x25 display.  But I always think of 'em as 80x24, and I think pretty much all of them were that.  After around 1980, you started seeing a lot of terminals that supported longer lines, like 135 in the Data General D460.  But I think it's better to stick with 80x24, because pretty much everything will support that.

Ironically, many versions of BASIC in the 1970s - which most of these micros ran - had a 72 character width limitation.  But that doesn't matter for what we're doing here.  The terminals connected to the system were 80x24, so the program output can take advantage of that.

As for escape sequences, seems like everything after around 1978 used escape sequences in addition to control characters.  The escape sequences were of varying length, but I don't recall any that were longer than four or five characters.  You pretty much just have to download the manuals for the terminals you want to support, and map their escape sequences to the function you want to perform.

I'd love to see your games for the Sym-1!  Let me know when you have some, and I'd be happy to test them for you!

This is what I've done, just a handful of little things like a version of Conway's Life, Mandelbrot, etc.

http://ParhamData.com/Sym1/C_applications_source.zip

Fabrizio-Caruso commented:
@WayneParham thanks for your reply!
Which terminal do you still use? If I add support for the SYM-1 in my library, and, as I fear, no emulator provides a terminal display, I will ask you to try my games on the real thing... So I need to know what you still use.

By the way, if you curious, you can take a look at my universal library/framework/abstraction layer:
https://github.com/Fabrizio-Caruso/CROSS-LIB/

WayneParham commented:
Your CrossLib looks great!  And five games for it already!  I'll have to give it a more detailed look as time permits.

I use a VT-100 right now.  I've been kicking myself 'cause I had about 20 Lear Sigler ADM3A terminals and as many Data General D200 terminals in storage up to the mid 1990s.  After that, I "realized" I was paying more in storage each month than they were worth, so I got rid of them all.  Today, they're worth about a thousand dollars each.

Years passed.

When I "dusted off" my Sym-1 in 2020, I powered it up and connected using Putty.  That has a VT-100 emulator built-in.  But I wanted a dedicated terminal, so I looked for one on eBay.  That's when I learned dumb terminals from the 1970s were selling for a grand these days.  Yikes!

I found a really cool little project though.  Check out the "VT-100ish" I'm running now!

http://ParhamData.com/VT100ish
